### ISSUE TITLE
Handling of `None` in torch.utils._pytree is inconsistent with JAX.

### ISSUE THREAD

**Issue by keunhong** [pytorch/pytorch#119328]
### 🐛 Describe the bug

`None`  is handled as a non-leaf in JAX by default, but it is treated as a leaf in the `torch.utils._pytree`. This is configurable in `optree` which has a default behavior consistent with JAX.

Treating `None` as a middle node is important because even simple examples like this quickly become cumbersome otherwise:
```
>> torch.utils._pytree.tree_map(lambda x: x.shape, {'a': None, 'b': torch.zeros(10)})
```
This currently results in an error:
```
AttributeError                            Traceback (most recent call last)
Cell In[9], line 1
----> 1 torch.utils._pytree.tree_map(lambda x: x.shape, {'a': None, 'b': torch.zeros(10)})

File ~/src/torchcam/.venv/lib/python3.11/site-packages/torch/utils/_pytree.py:602, in tree_map(func, tree)
    600 def tree_map(func: Callable[..., Any], tree: PyTree) -> PyTree:
    601     flat_args, spec = tree_flatten(tree)
--> 602     return tree_unflatten([func(i) for i in flat_args], spec)

File ~/src/torchcam/.venv/lib/python3.11/site-packages/torch/utils/_pytree.py:602, in <listcomp>(.0)
    600 def tree_map(func: Callable[..., Any], tree: PyTree) -> PyTree:
    601     flat_args, spec = tree_flatten(tree)
--> 602     return tree_unflatten([func(i) for i in flat_args], spec)

Cell In[9], line 1, in <lambda>(x)
----> 1 torch.utils._pytree.tree_map(lambda x: x.shape, {'a': None, 'b': torch.zeros(10)})

AttributeError: 'NoneType' object has no attribute 'shape'
```

This works fine in JAX:
```
>> jax.tree_util.tree_map(lambda x: x.shape, {'a': None, 'b': jax.numpy.zeros(10)})
{'a': None, 'b': (10,)}

```

### Versions

Collecting environment information...
PyTorch version: 2.2.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.2.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.1.0.2.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.7 (main, Jan 19 2024, 18:31:00) [Clang 15.0.0 (clang-1500.1.0.2.5)] (64-bit runtime)
Python platform: macOS-14.2.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.3
[pip3] optree==0.10.0
[pip3] torch==2.2.0
[pip3] torchvision==0.17.0
[conda] No relevant packages

**Comment by suo** [comment]
Yeah this is a difference, but will probably be hard to change now that we have a lot of code depending on this behavior (cc @zou3519 who would know if there has been previous discussion on this).

In the meantime, `tree_map_only` might help you in this particular case

**Comment by keunhong** [comment]
Ah, thanks for the pointer. `tree_map_only` is certainly convenient for this specific case. I think a bigger issue is when using `torch.vmap` and friends that currently have this behavior as well (since it calls pytree under the hood), and as far as I can tell there's no way to have it call `tree_map_only`.

**Comment by zou3519** [comment]
@keunhong a workaround is to replace the None with []

**Comment by keunhong** [comment]
One of our data structures depends on allowing to set certain fields to `None`, and the logic switches based on that.

Having any `None` values in the pytree currently is rejected by `torch.vmap`:

```
def f(x):
    return x


torch.vmap(f)({'a': None, 'b': torch.zeros((10, 3))})
```

We could change `None` to `[]` like you suggested, but this would mean changing all of the logic to check for lists, and would also preclude the data structure from containing lists.

**Comment by zou3519** [comment]
Hmm, maybe we could have an option (a flag to vmap) for this. But we would need to add the option to our pytree implementation

**Comment by keunhong** [comment]
Here's what I've tried to far hoping to find a workaround, but no luck yet:
* I tried to add `None` to as a node type by calling `register_pytree_node` but it looks like there are some checks that disallow registering non-class objects.
* I tried monkey-patching the pytorch pytree to point to the optree pytree instead, but it looks like there are some checks within vmap itself that block this usecase.

**Comment by keunhong** [comment]
I've temporarily worked around this by defining a sentinel 'DisabledField' dataclass but it is definitely not ideal since we need to remember to use this instead of None everywhere. Do you think adding the option os a possibility?

**Comment by zou3519** [comment]
@XuehaiPan thoughts?

**Comment by XuehaiPan** [comment]
> Hmm, maybe we could have an option (a flag to vmap) for this. But we would need to add the option to our pytree implementation

Maybe we can add the extra option `none_is_leaf` (defaults to `True` for BC) like `optree` do. This is more flexible and keeps backward compatibility.

**Comment by XuehaiPan** [comment]
Another approach is to provide a context manager to temporarily enable/disable types in the pytree node registry.

For example, we have a class `_PyTreeExtensionContext`:

https://github.com/pytorch/pytorch/blob/de600508018ee208dd8f65532d53bd59e14f85aa/torch/onnx/_internal/fx/dynamo_graph_extractor.py#L32-L53

Example usage:

```python
with (
    PyTreeContext(...)
    .register(type(None), ...)
    .register(SomeType, ...)
    .unregister(SomeAlreadlyRegisteredType, ...)
):
    leaves, spec = tree_flatten(tree)
    other_tree = tree_unflatten(other_leaves, spec)

    mapped = tree_map(fn, tree)

# Potential misuse
spec  # !!! should not access the treespec created in the context after exiting the context !!!
```

**Comment by suo** [comment]
I would note that any flag that alters the runtime behavior of pytree handling will need to be accounted for in the serialization system. For example, if someone serialized with `none_is_leaf=True`, then what should happen if you deserialize with it set to False?

**Comment by XuehaiPan** [comment]
> I would note that any flag that alters the runtime behavior of pytree handling will need to be accounted for in the serialization system.

Agree. We should have a default fallback path when users try to load data which serialized with older version of PyTorch.

> For example, if someone serialized with `none_is_leaf=True`, then what should happen if you deserialize with it set to False?

The `none_is_leaf` flag should be part of the serialized representation. The deserialization function should not have the flag as function argument.

**Comment by keunhong** [comment]
Wanted to bump this to see if there were any plans to add this capability. We've been using the following workaround but it feels so ugly and I'm eager to get rid of it.

```python
import optree
from torch.utils import _pytree

class NoneField:
    """
    A sentinel value that marks a field as disabled when set to this value.
    """

    def __eq__(self, other):
        return isinstance(other, NoneField)


def is_none(value: Any) -> bool:
    """Return True if the field is disabled."""
    return isinstance(value, NoneField) or value is None


optree.register_pytree_node(NoneField, lambda _: ([], []), lambda *_: NoneField())
_pytree.register_pytree_node(
    NoneField,
    lambda _: ([], []),
    lambda *_: NoneField(),
    flatten_with_keys_fn=lambda _: ([], []),
)
```

**Comment by XuehaiPan** [comment]
The solution in the PyTorch pytree way is:

```python
torch_pytree.tree_map_only(torch.Tensor, func, tree)  # only call func on torch.Tensor instances
```

**Comment by keunhong** [comment]
It appears that `tree_map_only` is missing the `*rest` argument for tree multimaps. If it had that it could be a drop-in replacement like this:
```
from torch.utils import _pytree

_pytree.tree_map_only(
    lambda x: x is not None, 
    lambda x, y: x + y, 
    {"a": 1, "c": torch.tensor(2), "b": None},
     {"a": 1, "c": torch.tensor(2), "b": None},

```

**Comment by XuehaiPan** [comment]
> What I'd like to do is to have a local version of `tree_map` that replicates the JAX behavior:

@keunhong The correct JAX replicate is:

```python
def tree_map(
    func: Callable[..., Any],
    tree: PyTree,
    *rests: PyTree,
    is_leaf: Optional[Callable[[PyTree], bool]] = None,
) -> PyTree:
    def func_skip_none(*args):
        if args[0] is None:
            if not all(x is None for x in args):
                raise ValueError
            return None
        return func(*args)

    return torch_pytree.tree_map(
        func_skip_none,
        tree,
        *rests,
        is_leaf=is_leaf,
    )
```

**Comment by keunhong** [comment]
Ah yes of course, that works! It's been a while since I've thought about this.

I think the main issue after this is that a lot of internal functions like `torch.vmap` use pytree internally but I'm not sure there's a way to override the tree functions there.

**Comment by XuehaiPan** [comment]
> I think the main issue after this is that a lot of internal functions like `torch.vmap` use pytree internally 

Maybe this should be discussed in a new issue for `vmap` with a repro script.

> but I'm not sure there's a way to override the tree functions there.

Changing how we handle `None` is a BC-breaking change.

**Comment by keunhong** [comment]
Hi, I wanted to follow up on this issue with a concrete use case and a potential feature request.

## Our Use Case

We have PyTree dataclasses with optional fields that may be `None`. We'd like `tree_map` and `vmap` to skip these fields when they're absent, similar to how JAX handles `None` (as a non-leaf that's ignored during traversal).

## Current Workaround

We created a `NoneField` sentinel class registered with no children:

```python
class NoneField:
    def __bool__(self):
        return False  # Falsy like None

    def __eq__(self, other):
        return isinstance(other, NoneField) or other is None

_pytree.register_pytree_node(
    NoneField,
    lambda _: ([], []),  # No children
    lambda *_: NoneField(),
)
```

This works, but requires users to check `if not field:` instead of `if field is None:`.

## Why We Can't Register `type(None)` Globally

We initially tried registering `type(None)` with no children:

```python
_pytree.register_pytree_node(
    type(None),
    lambda _: ([], []),
    lambda *_: None,
)
```

This breaks `torch.vmap` because `None` in `in_dims` is a special marker meaning "don't batch this argument":

```python
# Without registration:
tree_flatten((0, None)) -> ([0, None], spec)

# With registration:
tree_flatten((0, None)) -> ([0], spec)  # None is "flattened away"

# This breaks vmap:
vmap(fn, in_dims=(0, None))(batched_tensor, unbatched_tensor)
# Now in_dims structure doesn't match the inputs
```

## Feature Request

Would it be possible to decouple the pytree handling of `None` from `vmap`'s special `in_dims` semantics?

Potential approaches:
1. Have `vmap` handle `in_dims` separately before pytree flattening
2. Add a flag to `register_pytree_node` that excludes certain contexts (like `in_dims`)
3. Use a different sentinel value for "don't batch" in `in_dims` (though this would be a breaking change)

The ideal outcome would be that users can register `type(None)` to have no children (matching JAX behavior) without breaking `vmap`'s `in_dims` functionality.
